src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs

```diff
@@ -165,26 +167,17 @@ SymbolKind.NetModule or
 
                         var newState = CompilationTrackerState.Create(
                             withCompilationState.CompilationWithoutGeneratedDocuments,
-                            state.GeneratorInfo,
+                            withCompilationState.GeneratorInfo,
                             finalCompilationWithGeneratedDocuments,
                             intermediateProjects);
                         return newState;
                     }
                     else if (state is NoCompilationState)
                     {
-                        // We may still have a cached generator; we'll have to remember to run generators again since we are making some
-                        // change here. We'll also need to update the other state of the driver if appropriate.
-
-                        // The no compilation state can never be in the 'DocumentsAreFinal' state.  The only place where
-                        // we start with the NoCompilationState is the 'Empty' instance (where DocumentsAreFinal=false).
-                        // And then this is the only place where we get a NoCompilationState and create a new instance.
-                        // So there is no way to ever transition this to the DocumentsAreFinal=true state.
-                        Contract.ThrowIfTrue(state.GeneratorInfo.DocumentsAreFinal);
-                        var generatorInfo = state.GeneratorInfo;
-                        if (generatorInfo.Driver != null && translate != null)
-                            generatorInfo = generatorInfo.WithDriver(translate.TransformGeneratorDriver(generatorInfo.Driver));
-
-                        return new NoCompilationState(generatorInfo);
+                        // This is a singleton.  We just stay in this state.  That's fine as there's no driver or
+                        // compilation to track this translate step.  We'll always just have to produce the full
+                        // compilation/sg docs if we're in this state.
+                        return state;
```

Comment:
had offline chat where i explained my reasoning.  First, there are only two ways to get a NoCompilationState.  Either because we were at the 'initial empty' state, or because we got into this codepath (we don't create a NoCompilationState *anywhere* else).  

But this codepath checks if .Driver is null or not.  But if you have the initial empty state, it is null.  And as such, we never update the generator info.  So the generator info *always* stays in the default state (it never get documents, and never changes the 'final' state, and never changse teh 'Driver' state).  So we literally never mutate anything here.

This is backed up with this pr: https://github.com/dotnet/roslyn/pull/71992

So i feel like this is call simplified by moving this to be this singleton.

Note: we could even consider 'null' to be the singleton.

Reply:
This matches my comment on the other PR: this could have happened when compilations could get GCed, but that's not the case anymore.

Reply:
yup!